packages/model-viewer/src/three-components/gltf-instance/VariantMaterialLoaderPlugin.ts

```diff
@@ -0,0 +1,275 @@
+/**
+ * Materials variants extension
+ *
+ * Specification:
+ * https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_variants
+ */
+
+import {Material, Mesh, Object3D} from 'three';
+import {GLTF, GLTFLoaderPlugin, GLTFParser} from 'three/examples/jsm/loaders/GLTFLoader.js';
+
+/**
+ * KHR_materials_variants specification allows duplicated variant names
+ * but it makes handling the extension complex.
+ * We ensure tha names and make it easier.
+ * If you want to export the extension with the original names
+ * you are recommended to write GLTFExporter plugin to restore the names.
+ *
+ * @param variantNames {Array<string>}
+ * @return {Array<string>}
+ */
+const ensureUniqueNames = (variantNames: string[]) => {
+  const uniqueNames = [];
+  const knownNames: {[key: string]: boolean} = {};
+
+  for (const name of variantNames) {
+    let uniqueName = name;
+    let suffix = 0;
+    // @TODO: An easy solution.
+    //        O(N^2) in the worst scenario where N is variantNames.length.
+    //        Fix me if needed.
+    while (knownNames[uniqueName] !== undefined) {
+      uniqueName = name + '.' + (++suffix);
+    }
+    knownNames[uniqueName] = true;
+    uniqueNames.push(uniqueName);
+  }
+
+  return uniqueNames;
+};
+
+/**
+ * Convert mappings array to table object to make handling the extension easier.
+ *
+ * @param
+ *     extensionDef {glTF.meshes[n].primitive.extensions.KHR_materials_variants}
+ * @param variantNames {Array<string>} Required to be unique names
+ * @return {Object}
+ */
+const mappingsArrayToTable = (extensionDef: any, variantNames: string[]) => {
+  const table:
+      {[key: string]:
+           {material: Material|null, gltfMaterialIndex: number}} = {};

```

Comment:
Can we use `table = new Map<string, {material: Material|null, gltfMaterialIndex: number}>();` instead? Basically the same, but a bit more modern.

Reply:
Done